api/v1alpha1/bundle_types.go

```diff
@@ -17,34 +17,65 @@ limitations under the License.
 package v1alpha1
 
 import (
+	corev1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
-type BundleReference string
+type BundleSourceType string
+
+// BundleVolumeMount is the specification of a single
+// volume that contains bundle manifest content and
+// how to mount that content.
+type BundleVolumeMount struct {
+	// MountPath is the filesystem path to the bundle
+	// manifest content
+	MountPath string `json:"mountPath"`
+}
+
+// +union
+type BundleSource struct {
+	// +unionDiscriminator
+	// +optional
+	VolumeMounts []BundleVolumeMount `json:"volumeMounts,omitempty"`
+	Ref          string              `json:"ref,omitempty"`
+}
 
 // BundleSpec defines the desired state of Bundle
 type BundleSpec struct {
-	// Class specifies the name of the ProvisionerClass to use for unpacking the bundle.
-	Class string `json:"class,omitempty"`
+	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
+	// Important: Run "make" to regenerate code after modifying this file
 
-	Refs []BundleReference `json:"source"`
+	// Likely in the future, we'll have the immediate need to handle something
+	// like the `olm.bundle` class which has embedded knowledge that understands
+	// how to unpack registry+v1 bundle image(s).
+	// Class specifies the name of the ProvisionerClass to use for unpacking the bundle.
+	Class  ProvisionerID `json:"class,omitempty"`
+	Source BundleSource  `json:"source"`
 }
 
 // BundleStatus defines the observed state of Bundle
 type BundleStatus struct {
-	// INSERT ADDITIONAL STATUS FIELD - define observed state of cluster
-	// Important: Run "make" to regenerate code after modifying this file
-	Conditions []metav1.Condition `json:"conditions,omitempty"`
-
-	// TODO: Reference to volume containing unpacked bundle content
+	Conditions []metav1.Condition           `json:"conditions,omitempty"`
+	Unpacked   BundleUnpackStatusType       `json:"unpacked,omitempty"`
+	Volume     *corev1.LocalObjectReference `json:"volume"`
+	URI        string                       `json:"uri,omitempty"`
 }
 
+type BundleUnpackStatusType string
+
+const (
+	BundleUnpacked       BundleUnpackStatusType = "Unpacked"
```

Comment:
I think these status types might be overly specified to the existing olm bundle type -- for example, if the bundle API is used for a helm chart or a git repository, would we call pulling them from remote to a PV on-cluster "unpacking" in the same way as we do for olm based bundles? I'm wondering if there are some more generic terms that we could use to describe this behavior. 

We have "unpack" a bundle versus "install" a helmchart versus "clone" a repository essentially. 

Maybe we could use the term `Contents` as the top level field on the Bundle.Status and then the term `Available` to describe that a bundle has been fully unpacked and is available on cluster. `InProgress` to mean that the content unpacking is in progress. And `NotAvailable` by default when the bundle is created.

Reply:
Yeah completely agree - as an aside, everything related to how these status types are handled is up for grabs imo. I definitely like the idea of displaying availability, or approval, maybe these can be custom columns when querying local through kubectl would be a decent UX?